pulp_2to3_migration/app/plugin/content.py

```diff
@@ -198,62 +202,93 @@ def get_remote_by_importer_id(importer_id):
+            else:
+                # create artifact for content that has file
+                # TODO potentially improve this part for content that can have multiple files
+                artifact = await self.create_artifact(pulp2content.pulp2_storage_path,
+                                                      pulp_2to3_detail_content.expected_digests,
+                                                      pulp_2to3_detail_content.expected_size,
+                                                      downloaded=pulp2content.downloaded)
+
+                if lazy_type and pulp2lazycatalog:
+                    # handle DA and RA creation for content that supports on_demand
+                    # Downloaded or on_demand content with LCEs.
+                    #
+                    # To create multiple remote artifacts, create multiple instances of declarative
+                    # content which will differ by url/remote in their declarative artifacts
+                    for lce in pulp2lazycatalog:
+                        remote = get_remote_by_importer_id(lce.pulp2_importer_id)
+                        deferred_download = not pulp2content.downloaded
+                        if not remote and deferred_download:
+                            _logger.warn(_('On_demand content cannot be migrated without a remote '
+                                           'pulp2 unit_id: {}'.format(pulp2content.pulp2_id)))
+                            continue
+
+                        relative_path = pulp_2to3_detail_content.relative_path_for_content_artifact
+                        da = DeclarativeArtifact(
+                            artifact=artifact,
+                            url=lce.pulp2_url,
+                            relative_path=relative_path,
+                            remote=remote,
+                            deferred_download=deferred_download)
+                        dc = DeclarativeContent(content=pulp3content, d_artifacts=[da])
+                        dc.extra_data = future_relations
+                        await self.put(dc)
+                else:
+                    da = DeclarativeArtifact(
+                        artifact=artifact,
+                        url=NOT_USED,
+                        relative_path=pulp_2to3_detail_content.relative_path_for_content_artifact,
+                        remote=None,
+                        deferred_download=False)
+                    dc = DeclarativeContent(content=pulp3content, d_artifacts=[da])
+                    dc.extra_data = future_relations
+                    await self.put(dc)
 
             if pb:
                 pb.increment()
 
+            if has_future:
+                futures.append(dc)
+                resolve_futures = len(futures) >= batch_size or pb.done == pb.total
+                if resolve_futures:
+                    for dc in futures:
+                        await dc.resolution()
+                    futures.clear()
```

Comment:
What is the reason/benefit to accumulate futures and then process them in another loop separately, at the end of the batch processing?
Naive question. Is it bad to do:
```
if has_future:
    await dc.resolution() 
```
?

Reply:
it is better to send more items into the pipeline and await for it resolution later.
this way we will be able to process 500 rpms into each stage; first we are sending 500 pieces with self.put(), accumulate the dc in the list and then await for resolution for all of those,  otherwise if doing:
```
if has_future:
    await dc.resolution()
 ```

it will be 1 rpm in each stage